src/mongo_odm/query_builder.hpp

```diff
@@ -0,0 +1,466 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <mongo_odm/config/prelude.hpp>
+
+#include <cstddef>
+#include <stdexcept>
+#include <tuple>
+#include <type_traits>
+
+#include <bsoncxx/builder/basic/array.hpp>
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/basic/kvp.hpp>
+#include <bsoncxx/builder/basic/sub_array.hpp>
+#include <bsoncxx/builder/basic/sub_document.hpp>
+#include <bsoncxx/builder/core.hpp>
+#include <bsoncxx/view_or_value.hpp>
+
+#include <bsoncxx/builder/stream/array.hpp>
+
+#define NVP(x) mongo_odm::makeNvp(&mongo_odm_wrap_base::x, #x)
+
+#define ODM_MAKE_KEYS(Base, ...)      \
+    using mongo_odm_wrap_base = Base; \
+    constexpr static auto fields = std::make_tuple(__VA_ARGS__);
+
+#define ODM_MAKE_KEYS_STORAGE(Base) constexpr decltype(Base::fields) Base::fields
```

Comment:
I actually did an update to my initial example that should get rid of this constraint.

https://github.com/hanumantmk/field_member_magic/commit/86c6436754e395be9c1eeaba5cf281c64c62396d

If you make fields a function which returns the tuple, you don't need storage